Cython/Utility/Coroutine.c

```diff
@@ -1537,38 +1537,64 @@ __Pyx_Coroutine_set_qualname(__pyx_CoroutineObject *self, PyObject *value, void
     return 0;
 }
 
+// called from in a critical section on self
 static PyObject *
-__Pyx_Coroutine_get_frame(__pyx_CoroutineObject *self, void *context)
+__Pyx__Coroutine_get_frame(__pyx_CoroutineObject *self)
 {
 #if !CYTHON_COMPILING_IN_LIMITED_API
     PyObject *frame = self->gi_frame;
-    CYTHON_UNUSED_VAR(context);
     if (!frame) {
         if (unlikely(!self->gi_code)) {
             // Avoid doing something stupid, e.g. during garbage collection.
             Py_RETURN_NONE;
         }
+        // The coroutine doesn't know what module it's in so can't fill in proper globals.
+        // In principle this could be solved by storing a reference to the module in the coroutine,
+        // but in practice it probably isn't worth the extra memory.
```

Comment:
To me, it doesn't sound unreasonable to keep a reference to at least the module globals (or the module itself?) also in the coroutine. We do that for functions, why not for coroutines? In fact, seeing us move more towards sharing type implementations between multiple modules, it seems rather obvious to me that their instances should know where they came from.